src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,682 @@
+
+        let addr = builder.spawn(sink, sink_context)?;
+        Ok(Some(addr))
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
+    data: Vec<u8>,
+    block_size: usize,
+    buffer_start: usize,
+}
+
+impl ChunkedBuffer {
+    pub fn new(size: usize) -> Self {
+        Self {
+            data: Vec::with_capacity(size * 2),
+            block_size: size,
+            buffer_start: 0,
+        }
+    }
+
+    pub fn mark_done_until(&mut self, position: usize) -> Result<()> {
+        if position < self.buffer_start {
+            return Err("Buffer was marked as done at index which is not in memory anymore".into());
+        }
+
+        let bytes_to_remove = position - self.buffer_start;
+        self.data = Vec::from(self.data.get(bytes_to_remove..).ok_or(
+            ErrorKind::GoogleCloudStorageError("Not enough data in the buffer"),
+        )?);
+        self.buffer_start += bytes_to_remove;
+
+        Ok(())
+    }
+
+    pub fn read_current_block(&self) -> Option<&[u8]> {
+        self.data.get(..self.block_size)
+    }
+
+    pub fn write(&mut self, data: &[u8]) {
+        self.data.extend_from_slice(data);
+    }
+
+    pub fn start(&self) -> usize {
+        self.buffer_start
+    }
+
+    pub fn end(&self) -> usize {
+        self.buffer_start + self.data.len().min(self.block_size)
+    }
+
+    pub fn final_block(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+struct GCSWriterSink {
+    #[cfg(test)]
+    client: Option<tests::MockHttpClient>,
+    #[cfg(not(test))]
+    client: Option<H1Client>,
+    url: Url<HttpsDefaults>,
+    config: Config,
+    buffers: ChunkedBuffer,
+    current_name: Option<String>,
+    current_session_url: Option<String>,
+}
+
+#[async_trait::async_trait]
+impl Sink for GCSWriterSink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        for (value, meta) in event.value_meta_iter() {
+            let meta = ctx.extract_meta(meta);
+
+            let name = meta
+                .get("name")
+                .ok_or(ErrorKind::GoogleCloudStorageError(
+                    "Metadata is missing the file name",
+                ))?
+                .as_str()
+                .ok_or(ErrorKind::GoogleCloudStorageError(
+                    "The file name in metadata is not a string",
+                ))?;
+
+            self.finish_upload_if_needed(name).await?;
+
+            self.start_upload_if_needed(meta, name).await?;
+
+            let serialized_data = serializer.serialize(value, event.ingest_ns)?;
+            for item in serialized_data {
+                self.buffers.write(&item);
+            }
+
+            if let Some(data) = self.buffers.read_current_block() {
+                let mut response = None;
+                for i in 0..3 {
```

Comment:
Do we really want to keep this retry logic? We don't do similar things e.g. with the `http_client` connector. If people want retries, they can put a WAL in between or make their events `transactional` with other means. What do you think?

Reply:
We could remove it, though in my experience, the errors here happen quite often.

Reply:
The one big problem i have with this is that if we sleep here waiting for a retry, we are blocking the whole `sink` task from making progress. The http_client spawns a task for every http request sent. Could we somehow do the same here, keep track of the tasks/joinhandles/futures and chain those futures, so the uploads actually happen in order?

Reply:
hmm... that could work, we could also have a background task that deals with the communication with GCS, but this brings another issue - how do we deal with ACKs then? AFAIK the ACK has to happen as `on_event` returns, unless I'm missing something

Reply:
Here is an example from the http_client: 
* https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/http/client.rs#L164
* https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/http/client.rs#L373-L378

Reply:
Since we have the retry logic in 3 places here, would it make sense to abstract it out, make it configurable and allow it be used elsewhere too?

hardcoded 3 tries at 25ms^2*i seems like a recepie for pain :joy:

Reply:
Yep, all good points, I'll revamp the error handling and retries.